README.md
-------------------------------------------
DRAKMA - HTTP client written in Common Lisp
-------------------------------------------

DRAKMA is a HTTP client written in Common Lisp.  Please visit [the
documentation site](https://edicl.github.io/drakma/) for more information.

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/edicl/drakma?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)


CHANGELOG
Version 2.0.7
2019-10-23
Don't pass :TLSEXT-HOST-NAME on older LispWorks.

Version 2.0.6
2019-09-27
Pass :TLSEXT-HOST-NAME option when ATTACH-SSL on LispWorks (Xiangyu He)

Version 2.0.5
2019-04-07

Version 2.0.4
2017-08-28
Support all :VERIFY options with cl+ssl. (Anton Vodonosov)

Version 2.0.3
2016-12-19
Allegro CL Express SSL fix (Chris Riesbeck)
Use force-ssl, verify and other SSL settings when using proxy (lsxvdqe)
verify, ca-path, ca-directory and max-depth for CL+SSL (lsxvdqe)
Using host from :additional-headers when specified (Azamat S. Kalimoulline)

Version 2.0.2
2015-10-08
Support SNI with cl+ssl. (Paul M. Rodriguez)

Version 2.0.1
2015-07-10
Revert "Use quri instead of puri" (Hans Hübner)
Merge pull request #56 from tmccombs/test-op (Hans Hübner)
Add test-op to ASDF for drakma and drakma-test systems. (Thayne McCombs)

Version 2.0.0
2015-07-10
Use quri instead of puri for URIs. (Thayne McCombs)

Version 1.3.15
2015-07-07
Refactor time zone parsing into regular expression (Hans Huebner)
Update README.md (Hans Hübner)
make-ssl-stream: declare additional variables ignorable (Mark David)

Version 1.3.14
2015-03-14
Update documentation (Hans Huebner)
Add DECODE-CONTENT option (Thayne McCombs)

Version 1.3.13
2015-03-11
Workaround for ABCL streams bug More information here: http://abcl.org/trac/ticket/377 (Elias Martenson)
Fix code to generate version number in HTML docs (Hans Huebner)

Version 1.3.12
2015-01-15
Preserve octet element-type with :force-binary t. (Jan Moringen)

Version 1.3.11
2014-11-28
Update support info (Hans Huebner)
Do not escape URI twice. (Orivej Desh)

Version 1.3.10
2014-09-16
Implement gzip and deflate decoding for responses. (Thayne McCombs)
Fix error in example (reported by David Vázquez Púa, closes #44) (Hans Huebner)

Version 1.3.9
2014-05-01
Added SSL/TLS support for mocl (Wukix Inc)
reindent and refactor cond into if (Hans Huebner)

Version 1.3.8
2014-02-25
Determine version number from asdf package (Hans Huebner)
Silence warning on LispWorks (Hans Huebner)
All symbols from puri are prefixed with puri:. (Kilian Sprotte)

Version 1.3.7
2013-11-21
Dummy without any functional changes

Version 1.3.6
Documentation fixes (thanks to sarvid and stassats for the report)

Version 1.3.5
Ignore incoming Content-Length header when chunking is on
Make POST requests use external-format-out for multipart/form-data. (Raymond Wiker)
Fixed the link to the SBCL documentation pertaining to "defining constants" (Aaron France)
Added *NO-PROXY-DOMAINS* special variable (Aaron France)

Version 1.3.4
Add *default-http-proxy* special variable (Tomas Zellerin)

Version 1.3.3
Change handling of empty and missing Location headers (Paul M. Rodriguez)

Version 1.3.2
Redirect to GET only for POST requests (Vsevolod Dyomkin)

Version 1.3.1
2013-03-23
When redirecting from POST to GET, clear the FORM-DATA flag
Add trivial test suite (Anton Vodonosov)

Version 1.3.0
2012-12-28
Redirect HTTP 302 and 303 using GET (Orivej Desh)
Add URL-ENCODER keyword argument
Move documentation to XML format and update docstrings

Version 1.2.9
2012-10-18
Fix bug with Content-Length computation (Manabu Takayama)
Add REAL-HOST keyword argument (Orivej Desh)

Version 1.2.8
2012-09-12
fix the computation of request's Content-Length (Manabu Takayama)

Version 1.2.7
2012-08-16
Support :REPORT method (Cyrus Harmon)
Make PRESERVE-URI work better - PURI:URI mangles paths with encoded &'s.

Version 1.2.6
2012-03-03
Enable timeouts for more implementations (Francisco Vides Fernández)
Export URL-ENCODE (suggested by Rob Blackwell)
Fix incorrect range header syntax

Version 1.2.5
2012-01-30
use cl-ppcre:split instead of split-string to fix bug with GET
parameter handling (thanks to Rob Blackwell)
use :nodelay :if-supported (Anton Vodonosov)
Allow specification of client certificate (all platforms)
Add arguments that allow validation of server certificate

Version 1.2.4
2011-08-31
Make sure GET parameters are always URL-encoded
Add :RANGE keyword argument (Hans Huebner)
Better handling of optional filenames when uploading (Stas Boukarev)
Don't funcall symbols that aren't FBOUNDP (Fare Rideau)
Allow disabling of SSL when building (Marko Kocic)

Version 1.2.3
2010-08-05
Fix UPDATE-COOKIES (Vsevolod Dyomkin)
Fix typo in documentation HTML (Walter Rader)

Version 1.2.2
2010-07-10
Make sure pathless URIs work (Hans Huebner, Manuel Odendahl)

Version 1.2.1
2010-05-19
Fix a couple of typos (thanks to Stelian Ionescu, Giovanni Gigante, and Zach Beane)

Version 1.2.0
2010-05-19
Introduced *REMOVE-DUPLICATE-COOKIES-P* (Ryan Davis)
Enabled https through a proxy (Bill St. Clair and Dave Lambert)
Bugfix for redirect of a request through a proxy (Bill St. Clair)
Export PARSE-COOKIE-DATE
Safer method to render URIs
Allowed for GET/POST parameters without a value (seen on Lotus webservers)

Version 1.1.0
2009-12-01
Allowed additional headers to be function designators (suggested by Xiangjun Wu)
Be more liberal when parsing cookies (thanks to Andrei Stebakov)
Added HTTP method PATCH (thanks to Xiangjun Wu)
Don't send GET parameters again when redirecting (reported by Eugene Ossintsev)
Solidify feature expressions (thanks to Joshua Taylor)
Make SEND-COOKIE-P work for pathless URIs (thanks to Tomo Matsumoto)

Version 1.0.0
2009-02-19
Use the new ("binary") version of Chunga
Added conditions types
Some performance improvements
Be more lenient about content length (thanks to Zach Beane and "pix")
Added *ALLOW-DOTLESS-COOKIE-DOMAINS-P* (thanks to Daniel Janus)
Fix generation of user agent header (bug caught by Chaitanya Gupta)
Added DEADLINE parameter for CCL (thanks to Hans Huebner)
Fixed bug where READ-BODY returned NIL although TEXTP was true (thanks to Hans Huebner)

Version 0.11.5
2008-03-21
Added workaround for CLISP (thanks to Anton Vodonosov)

Version 0.11.4
2008-02-13
Improved error detection in MAKE-FORM-DATA-FUNCTION (suggested by Daniel Janus)

Version 0.11.3
2008-01-14
The previous change is only needed for Windows


test/drakma-test.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-

;;; Copyright (c) 2013, Anton Vodonosov.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(defpackage :drakma-test
  (:use :cl :fiveam))

(in-package :drakma-test)

(def-suite :drakma)
(in-suite :drakma)

(test get-google
  (let ((drakma:*header-stream* *standard-output*))
    (multiple-value-bind (body-or-stream status-code)
        (drakma:http-request "http://google.com/")
      (is (> (length body-or-stream) 0))
      (is (= 200 status-code)))))

(test get-google-ssl
  (let ((drakma:*header-stream* *standard-output*))
    (multiple-value-bind (body-or-stream status-code)
        (drakma:http-request "https://google.com/")
      (is (> (length body-or-stream) 0))
      (is (= 200 status-code)))))

(test post-google
  (let ((drakma:*header-stream* *standard-output*))
    (multiple-value-bind (body-or-stream status-code headers uri stream must-close reason-phrase)
        (drakma:http-request "http://google.com/" :method :post :parameters '(("a" . "b")))
      (declare (ignore headers uri stream must-close))
      (is (> (length body-or-stream) 0))
      (is (= 405 status-code))
      (is (string= "Method Not Allowed" reason-phrase)))))

(test post-google-ssl
  (let ((drakma:*header-stream* *standard-output*))
    (multiple-value-bind (body-or-stream status-code headers uri stream must-close reason-phrase)
        (drakma:http-request "https://google.com/" :method :post :parameters '(("a" . "b")))
      (declare (ignore headers uri stream must-close))
      (is (> (length body-or-stream) 0))
      (is (= 405 status-code))
      (is (string= "Method Not Allowed" reason-phrase)))))


(test gzip-content
  (let ((drakma:*header-stream* *standard-output*)
        (drakma:*text-content-types* (cons '(nil . "json") drakma:*text-content-types*)))
    (multiple-value-bind (body-or-stream status-code)
        (drakma:http-request "http://httpbin.org/gzip" :decode-content t)
      (is (= 200 status-code))
      (is (typep body-or-stream 'string))
      (is (search "\"gzipped\": true" body-or-stream)))))

(test deflate-content
  (let ((drakma:*header-stream* *standard-output*)
        (drakma:*text-content-types* (cons '(nil . "json") drakma:*text-content-types*)))
    (multiple-value-bind (body-or-stream status-code)
        (drakma:http-request "http://httpbin.org/deflate" :decode-content t)
      (is (= 200 status-code))
      (is (typep body-or-stream 'string))
      (is (search "\"deflated\": true" body-or-stream)))))

(test gzip-content-undecoded
      (let ((drakma:*header-stream* *standard-output*))
        (multiple-value-bind (body-or-stream status-code)
            (drakma:http-request "http://httpbin.org/gzip")
          (is (= 200 status-code))
          (is (typep body-or-stream '(vector flexi-streams:octet)))
          (is (> (length body-or-stream) 0))
          (is (equalp #(#x1f #x8b)
                      (subseq body-or-stream 0 2))))))

(test deflate-content-undecoded
      (let ((drakma:*header-stream* *standard-output*))
        (multiple-value-bind (body-or-stream status-code)
            (drakma:http-request "http://httpbin.org/deflate")
          (is (= 200 status-code))
          (is (typep body-or-stream '(vector flexi-streams:octet)))
          (is (> (length body-or-stream) 0))
          (is (equalp #x78 (aref body-or-stream 0))))))

(test stream
  (multiple-value-bind (stream status-code)
      (drakma:http-request "http://google.com/" :want-stream t)
    (is (streamp stream))
    (is (= 200 status-code))
    (is (subtypep (stream-element-type stream) 'character))
    (let ((buffer (make-string 1)))
      (read-sequence buffer stream))))

(test force-binary
  (multiple-value-bind (stream status-code)
      (drakma:http-request "http://google.com/" :want-stream t :force-binary t)
    (is (streamp stream))
    (is (= 200 status-code))
    (is (subtypep (stream-element-type stream) 'flexi-streams:octet))
    (let ((buffer (make-array 1 :element-type 'flexi-streams:octet)))
      (read-sequence buffer stream))))

(test verify.wrong.host
  (signals error
    (drakma:http-request "https://wrong.host.badssl.com/" :verify :required))
  (signals error
    (drakma:http-request "https://wrong.host.badssl.com/" :verify :optional))
  (finishes
    (drakma:http-request "https://wrong.host.badssl.com//" :verify nil)))

(test verify.expired
  (signals error
    (drakma:http-request "https://expired.badssl.com//" :verify :required))
  (signals error
    (drakma:http-request "https://expired.badssl.com/" :verify :optional))
  (finishes
    (drakma:http-request "https://expired.badssl.com/" :verify nil)))

(test verify.self-signed
  (signals error
    (drakma:http-request "https://self-signed.badssl.com/" :verify :required)))

(test verify.untrusted-root
  (signals error
    (drakma:http-request "https://untrusted-root.badssl.com/" :verify :required)))

(test verify.null
  (signals error
    (drakma:http-request "https://null.badssl.com/" :verify :required))
  (finishes
    (drakma:http-request "https://null.badssl.com/" :verify :optional)))



read.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: DRAKMA; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/drakma/read.lisp,v 1.17 2008/05/25 11:35:20 edi Exp $

;;; Copyright (c) 2006-2011, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :drakma)

(defun read-status-line (stream &optional log-stream)
  "Reads one line from STREAM \(using Chunga's READ-LINE*) and
interprets it as a HTTP status line.  Returns a list of two or
three values - the protocol \(HTTP version) as a keyword, the
status code as an integer, and optionally the reason phrase."
  (let* ((*current-error-message* "While reading status line:")
         (line (or (read-line* stream log-stream)
                   (error 'drakma-simple-error
                          :format-control "No status line - probably network error.")))
         (first-space-pos (or (position #\Space line :test #'char=)
                              (syntax-error "No space in status line ~S." line)))
         (second-space-pos (position #\Space line
                                     :test #'char=
                                     :start (1+ first-space-pos))))
    (list (cond ((string-equal line "HTTP/1.0" :end1 first-space-pos) :http/1.0)
                ((string-equal line "HTTP/1.1" :end1 first-space-pos) :http/1.1)
                (t (syntax-error "Unknown protocol in ~S." line)))
          (or (ignore-errors (parse-integer line
                                            :start (1+ first-space-pos)
                                            :end second-space-pos))
              (syntax-error "Status code in ~S is not an integer." line))
          (and second-space-pos (subseq line (1+ second-space-pos))))))

(defun get-content-type (headers)
  "Reads and parses a `Content-Type' header and returns it as
three values - the type, the subtype, and an alist \(possibly
empty) of name/value pairs for the optional parameters.  HEADERS
is supposed to be an alist of headers as returned by
HTTP-REQUEST.  Returns NIL if there is no such header amongst
HEADERS."
  (when-let (content-type (header-value :content-type headers))
    (with-sequence-from-string (stream content-type)
      (let* ((*current-error-message* "Corrupted Content-Type header:")
             (type (read-token stream))
             (subtype (and (assert-char stream #\/)
                           (read-token stream)))
             (parameters (read-name-value-pairs stream)))
        (values type subtype parameters)))))

(defun read-token-and-parameters (stream)
  "Reads and returns \(as a two-element list) from STREAM a token
and an optional list of parameters \(attribute/value pairs)
following the token."
  (skip-whitespace stream)
  (list (read-token stream)
        (read-name-value-pairs stream)))

(defun skip-more-commas (stream)
  "Reads and consumes from STREAM any number of commas and
whitespace.  Returns the following character or NIL in case of
END-OF-FILE."
  (loop while (eql (peek-char* stream nil) #\,)
        do (read-char* stream) (skip-whitespace stream))
  (skip-whitespace stream))

(defun read-tokens-and-parameters (string &key (value-required-p t))
  "Reads a comma-separated list of tokens from the string STRING.
Each token can be followed by an optional, semicolon-separated
list of attribute/value pairs where the attributes are tokens
followed by a #\\= character and a token or a quoted string.
Returned is a list where each element is either a string \(for a
simple token) or a cons of a string \(the token) and an alist
\(the attribute/value pairs).  If VALUE-REQUIRED-P is NIL, the
value part \(including the #\\= character) of each attribute/value
pair is optional."
  (with-sequence-from-string (stream string)
    (loop with *current-error-message* = (format nil "While parsing ~S:" string)
          for first = t then nil
          for next = (and (skip-whitespace stream)
                          (or first (assert-char stream #\,))
                          (skip-whitespace stream)
                          (skip-more-commas stream))
          for token = (and next (read-token stream))
          for parameters = (and token
                                (read-name-value-pairs stream
                                                       :value-required-p value-required-p))
          while token
          collect (if parameters (cons token parameters) token))))

(defun split-tokens (string)
  "Splits the string STRING into a list of substrings separated
by commas and optional whitespace.  Empty substrings are
ignored."
  (loop for old-position = -1 then position
        for position = (and old-position
                            (position #\, string :test #'char= :start (1+ old-position)))
        for substring = (and old-position
                             (trim-whitespace (subseq string (1+ old-position) position)))
        while old-position
        when (plusp (length substring))
        collect substring))


cookies.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: DRAKMA; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/drakma/cookies.lisp,v 1.15 2008/01/14 01:57:01 edi Exp $

;;; Copyright (c) 2006-2012, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :drakma)

(defclass cookie ()
  ((name :initarg :name
         :initform (cookie-error nil "A cookie must have a name.")
         :accessor cookie-name
         :documentation "The name of the cookie.")
   (value :initarg :value
          :initform ""
          :accessor cookie-value
          :documentation "The cookie's value.")
   (domain :initarg :domain
           :initform (cookie-error nil "A cookie must have a domain.")
           :accessor cookie-domain
           :documentation "The domain the cookie is valid for.")
   (path :initarg :path
         :initform "/"
         :accessor cookie-path
         :documentation "The path prefix the cookie is valid for.")
   (expires :initarg :expires
            :initform nil
            :accessor cookie-expires
            :documentation "When the cookie expires.  A Lisp
universal time or NIL.")
   (securep :initarg :securep
            :initform nil
            :accessor cookie-securep
            :documentation "Whether the cookie must only be
transmitted over secure connections.")
   (http-only-p :initarg :http-only-p
                :initform nil
                :accessor cookie-http-only-p
                :documentation "Whether the cookie should not be
accessible from Javascript.

This is a Microsoft extension that has been implemented in Firefox as
well. See <http://msdn2.microsoft.com/en-us/library/ms533046.aspx>."))
  (:documentation "Instances of this class represent HTTP cookies.  If
you need to create your own cookies, you should use MAKE-INSTANCE with
the initargs :NAME, :DOMAIN, :VALUE, :PATH, :EXPIRES,
:SECUREP, and :HTTP-ONLY-P all of which are optional except for the
first two.  The meaning of these initargs and the corresponding
accessors should be pretty clear if one looks at the original cookie
specification
<http://wp.netscape.com/newsref/std/cookie_spec.html> (and at this
page <http://msdn2.microsoft.com/en-us/library/ms533046.aspx> for the
HttpOnly extension)."))

(defun render-cookie-date (time)
  "Returns a string representation of the universal time TIME
which can be used for cookie headers."
  (multiple-value-bind (second minute hour date month year weekday)
      (decode-universal-time time 0)
    (format nil "~A, ~2,'0d-~2,'0d-~4,'0d ~2,'0d:~2,'0d:~2,'0d GMT"
            (aref #("Mon" "Tue" "Wed" "Thu" "Fri" "Sat" "Sun") weekday)
            date month year hour minute second)))

(defmethod print-object ((cookie cookie) stream)
  "Prints a representation of COOKIE similar to a `Set-Cookie' header."
  (print-unreadable-object (cookie stream :type t)
    (with-slots (name value expires path domain securep http-only-p)
        cookie
      (format stream "~A~@[=~A~]~@[; expires=~A~]~@[; path=~A~]~@[; domain=~A~]~@[; secure~]~@[; HttpOnly~]"
              name (and (plusp (length value)) value)
              (and expires (render-cookie-date expires))
              path domain securep http-only-p))))

(defun normalize-cookie-domain (domain)
  "Adds a dot at the beginning of the string DOMAIN unless there
is already one."
  (cond ((starts-with-p domain ".") domain)
        (t (format nil ".~A" domain))))

(defun valid-cookie-domain-p (domain)
  "Checks if the string DOMAIN contains enough dots to be
acceptable.  If *ALLOW-DOTLESS-COOKIE-DOMAINS-P* is non-NIL,
every domain name is considered acceptable."
  (or *allow-dotless-cookie-domains-p*
      (string-equal domain "localhost")
      (> (count #\. (normalize-cookie-domain domain) :test #'char=) 1)))

(defun cookie-domain-matches (domain uri)
  "Checks if the domain DOMAIN \(a string) matches the \(QURI) URI URI."
  (ends-with-p (normalize-cookie-domain (quri:uri-host uri))
               (normalize-cookie-domain domain)))

(defun send-cookie-p (cookie uri force-ssl)
  "Checks if the cookie COOKIE should be sent to the server
depending on the \(QURI) URI URI and the value of FORCE-SSL \(as
in HTTP-REQUEST)."
  (and ;; check domain
       (cookie-domain-matches (cookie-domain cookie) uri)
       ;; check path
       (starts-with-p (or (quri:uri-path uri) "/") (cookie-path cookie))
       ;; check expiry date
       (let ((expires (cookie-expires cookie)))
         (or (null expires)
             (> expires (get-universal-time))))
       ;; check if connection must be secure
       (or (null (cookie-securep cookie))
           force-ssl
           (eq (quri:uri-scheme uri) :https))))

(defun check-cookie (cookie)
  "Checks if the slots of the COOKIE object COOKIE have valid values
and raises a corresponding error of type COOKIE-ERROR otherwise."
  (with-slots (name value domain path expires)
      cookie
    (unless (and (stringp name) (plusp (length name)))
      (cookie-error cookie "Cookie name ~S must be a non-empty string." name))
    (unless (stringp value)
      (cookie-error cookie "Cookie value ~S must be a non-empty string." value))
    (unless (valid-cookie-domain-p domain)
      (cookie-error cookie "Invalid cookie domain ~S." domain))
    (unless (and (stringp path) (plusp (length path)))
      (cookie-error cookie "Cookie path ~S must be a non-empty string." path))
    (unless (or (null expires)
                (and (integerp expires)
                     (plusp expires)))
      (cookie-error cookie "Cookie expiry ~S should have been NIL or a universal time." expires))))

(defmethod initialize-instance :after ((cookie cookie) &rest initargs)
  "Check cookie validity after creation."
  (declare (ignore initargs))
  (check-cookie cookie))

(defmethod (setf cookie-name) :after (new-value (cookie cookie))
  "Check cookie validity after name change."
  (declare (ignore new-value))
  (check-cookie cookie))

(defmethod (setf cookie-value) :after (new-value (cookie cookie))
  "Check cookie validity after value change."
  (declare (ignore new-value))
  (check-cookie cookie))

(defmethod (setf cookie-domain) :after (new-value (cookie cookie))
  "Check cookie validity after domain change."
  (declare (ignore new-value))
  (check-cookie cookie))

(defmethod (setf cookie-path) :after (new-value (cookie cookie))
  "Check cookie validity after path change."
  (declare (ignore new-value))
  (check-cookie cookie))

(defmethod (setf cookie-expires) :after (new-value (cookie cookie))
  "Check cookie validity after expiry change."
  (declare (ignore new-value))
  (check-cookie cookie))

(defun cookie= (cookie1 cookie2)
  "Returns true if the cookies COOKIE1 and COOKIE2 are equal.
Two cookies are considered to be equal if name and path are
equal."
  (and (string= (cookie-name cookie1) (cookie-name cookie2))
       (string= (cookie-path cookie1) (cookie-path cookie2))))

(defclass cookie-jar ()
  ((cookies :initarg :cookies
            :initform nil
            :accessor cookie-jar-cookies
            :documentation "A list of the cookies in this cookie jar."))
  (:documentation "An object of this class encapsulates a
  collection (a list, actually) of COOKIE objects. You create a new
  cookie jar with (MAKE-INSTANCE 'COOKIE-JAR) where you can optionally
  provide a list of COOKIE objects with the :COOKIES initarg. The
  cookies in a cookie jar are accessed with COOKIE-JAR-COOKIES."))

(defmethod print-object ((cookie-jar cookie-jar) stream)
  "Print a cookie jar, showing the number of cookies it contains."
  (print-unreadable-object (cookie-jar stream :type t :identity t)
    (format stream "(with ~A cookie~:P)" (length (cookie-jar-cookies cookie-jar)))))

(defun parse-cookie-date (string)
  "Parses a cookie expiry date and returns it as a Lisp universal
time.  Currently understands the following formats:

  \"Wed, 06-Feb-2008 21:01:38 GMT\"
  \"Wed, 06-Feb-08 21:01:38 GMT\"
  \"Tue Feb 13 08:00:00 2007 GMT\"
  \"Wednesday, 07-February-2027 08:55:23 GMT\"
  \"Wed, 07-02-2017 10:34:45 GMT\"

Instead of \"GMT\" time zone abbreviations like \"CEST\" and UTC
offsets like \"GMT-01:30\" are also allowed.

While this function has \"cookie\" in its name, it might come in
handy in other situations as well and it is thus exported as a
convenience function.
"
  ;; it seems like everybody and their sister invents their own format
  ;; for this, so (as there's no real standard for it) we'll have to
  ;; make this function more flexible once we come across something
  ;; new; as an alternative we could use net-telent-date, but it also
  ;; fails to parse some of the stuff you encounter in the wild; or we
  ;; could try to employ CL-PPCRE, but that'd add a new dependency
  ;; without making this code much cleaner
  (handler-case
      (let* ((last-space-pos
              (or (position #\Space string :test #'char= :from-end t)
                  (cookie-date-parse-error "Can't parse cookie date ~S, no space found." string)))
             (time-zone-string (subseq string (1+ last-space-pos)))
             (time-zone (interpret-as-time-zone time-zone-string))
             second minute hour day month year)
        (dolist (part (rest (cl-ppcre:split "[ ,-]" (subseq string 0 last-space-pos))))
          (when (and day month)
            (cond ((every #'digit-char-p part)
                   (when year
                     (cookie-date-parse-error "Can't parse cookie date ~S, confused by ~S part."
                                              string part))
                   (setq year (parse-integer part)))
                  ((= (count #\: part :test #'char=) 2)
                   (let ((h-m-s (mapcar #'safe-parse-integer (cl-ppcre:split ":" part))))
                     (setq hour (first h-m-s)
                           minute (second h-m-s)
                           second (third h-m-s))))
                  (t (cookie-date-parse-error "Can't parse cookie date ~S, confused by ~S part."
                                              string part))))
          (cond ((null day)
                 (unless (setq day (safe-parse-integer part))
                   (setq month (interpret-as-month part))))
                ((null month)
                 (setq month (interpret-as-month part)))))
        (unless (and second minute hour day month year)
          (cookie-date-parse-error "Can't parse cookie date ~S, component missing." string))
        (when (< year 100)
          (setq year (+ year 2000)))
        (encode-universal-time second minute hour day month year time-zone))
    (cookie-date-parse-error (condition)
      (cond (*ignore-unparseable-cookie-dates-p*
             (drakma-warn "~A" condition)
             nil)
            (t (error condition))))))

(defun parse-set-cookie (string)
  "Parses the `Set-Cookie' header line STRING and returns a list
of three-element lists where each one contains the name of the
cookie, the value of the cookie, and an attribute/value list for
the optional cookie parameters."
  (let ((*current-error-message* (format nil "While parsing cookie header ~S:" string))
        result)
    (dolist (substring (split-set-cookie-string string))
      (with-sequence-from-string (stream substring)
        (let* ((name/value (read-name-value-pair stream :cookie-syntax t))
               (parameters (read-name-value-pairs stream :value-required-p nil :cookie-syntax t)))
          (push (list (car name/value) (cdr name/value) parameters) result))))
    (nreverse result)))

(defun get-cookies (headers uri)
  "Returns a list of COOKIE objects corresponding to the
`Set-Cookie' header as found in HEADERS \(an alist as returned by
HTTP-REQUEST).  Collects only cookies which match the domain of
the \(QURI) URI URI."
  (loop with set-cookie-header = (header-value :set-cookie headers)
        with parsed-cookies = (and set-cookie-header (parse-set-cookie set-cookie-header))
        for (name value parameters) in parsed-cookies
        for expires = (parameter-value "expires" parameters)
        for domain = (or (parameter-value "domain" parameters) (quri:uri-host uri))
        when (and (valid-cookie-domain-p domain)
                  (cookie-domain-matches domain uri))
        collect (make-instance 'cookie
                               :name name
                               :value value
                               :path (or (parameter-value "path" parameters)
                                         (quri:uri-path uri)
                                         "/")
                               :expires (and expires
                                             (plusp (length expires))
                                             (parse-cookie-date expires))
                               :domain domain
                               :securep (not (not (parameter-present-p "secure" parameters)))
                               :http-only-p (not (not (parameter-present-p "HttpOnly" parameters))))
		into new-cookies
	finally (return (ccase *remove-duplicate-cookies-p*
			  ((nil) new-cookies)
			  ((:keep-last t) (delete-duplicates new-cookies :test #'cookie=))
			  (:keep-first (delete-duplicates new-cookies :test #'cookie=
							  :from-end T))))))

(defun update-cookies (new-cookies cookie-jar)
  "Updates the cookies in COOKIE-JAR by replacing those which are
equal to a cookie in \(the list) NEW-COOKIES with the corresponding
`new' cookie and adding those which are really new."
  (setf (cookie-jar-cookies cookie-jar)
        (let ((updated-cookies
               (loop for old-cookie in (cookie-jar-cookies cookie-jar)
                     collect (or (find old-cookie new-cookies :test #'cookie=)
                                 old-cookie))))
          (union updated-cookies
                 (set-difference new-cookies updated-cookies :test #'cookie=)
                 :test #'cookie=)))
  cookie-jar)

(defun delete-old-cookies (cookie-jar)
  "Removes all cookies from COOKIE-JAR which have either expired
or which don't have an expiry date."
  (setf (cookie-jar-cookies cookie-jar)
        (loop with now = (get-universal-time)
              for cookie in (cookie-jar-cookies cookie-jar)
              for expires = (cookie-expires cookie)
              unless (or (null expires) (< expires now))
              collect cookie))
  cookie-jar)
